onnxscript/rewriter/ort_fusions/sdpa.py

```diff
@@ -41,173 +25,125 @@ def pattern(
+        _fusion_utils.check_shape(bindings, key_transposed, ["B", "H", "Dh", "Skv"])
+        _fusion_utils.check_shape(bindings, value, ["B", "H", "Skv", "Dv"])
+
+        def get_scale_value(tag_name: str, scale_name: str) -> float:
+            scaling_type = match_bindings.get(tag_name, "None")
+            if scaling_type == "None":
+                return 1.0
             else:
-                # Pass no scaling factor to SDPA, SDPA will use the default scaling factor
-                self._scale = None
+                scale = match_bindings.get(scale_name)
+                value = _ir_utils.get_singleton_value(scale)
+                if value is None:
+                    raise MatchFailureError(f"{scale_name} is not a scalar.", scale)
+                if scaling_type == "Mul":
+                    return value
+                else:
+                    assert scaling_type == "Div", f"Unexpected {scale_name} scaling operation"
+                    return 1.0 / value
 
-        # check ranks/shapes
+        query_scale_value = get_scale_value("query_scaling", "query_scale")
+        key_scale_value = get_scale_value("key_scaling", "key_scale")
+        qk_scale_value = get_scale_value("qk_scaling", "qk_scale")
+
+        self._scale = query_scale_value * key_scale_value * qk_scale_value
+
+        # If the scaling factor is the default one, we can skip passing it to SDPA.
+
+        head_size = bindings["Dh"]
+        if not isinstance(head_size, int):
+            return check_result
+
+        default_scaling_factor = 1.0 / math.sqrt(head_size)
+
+        if math.isclose(self._scale, default_scaling_factor, rel_tol=1e-5, abs_tol=1e-8):
+            # Pass no scaling factor to SDPA, SDPA will use the default scaling factor
+            self._scale = None
 
         return check_result
 
     def rewrite(
         self,
         op,
-        query,
-        key_transposed,
-        value,
-        mask,
-        query_scale,
-        key_scale,
-        qk_scale,
-        query_reshape=None,
+        query: ir.Value | None,
+        key_transposed: ir.Value | None,
+        value: ir.Value | None,
+        mask: ir.Value | None,
         **_,
     ):
-        if self._pre_scale and self._pre_scale_q:
-            if self._use_mul:
-                query_mul = op.Mul(query, qk_scale)
-            else:
-                query_mul = op.Div(query, qk_scale)
-            # Reshape and transpose 3D input of shape (B, S, D)
-            # to 4D input of shape (B, N, S, H)
-            if self._has_3d_query:
-                queryBNSH = op.Reshape(query_mul, query_reshape)
-                query = op.Transpose(queryBNSH, perm=[0, 2, 1, 3])
-            else:
-                query = query_mul
-
         sdpa_args = [query, key_transposed, value]
-        if self._use_mask:
+        if mask is not None:
             sdpa_args.append(mask)
         return op.SDPA(*sdpa_args, scale=self._scale, _domain="ai.onnxruntime.fusion")
```

Comment:
If self._scale is None, who decides the default, as op.SDPA is not a real op.

Reply:
I guess it is a matter of documenting SDPA. It is modeled closely on ONNX opset 23 Attention op. The default scale is common to most attention op variants, which is 1 / sqrt(headsize) as above.

Reply:
I am working on a follow-on PR that will implement SDPA using MHA. Hopefully that should make things more concrete too. (Will let us run all these test-cases with ORT after fusion.)